index.html

```diff
@@ -274,7 +274,7 @@ <h3>
 
           <pre class="example nohighlight" title="Creating the DID">
 did:web:w3c-ccg.github.io
- -> https://w3c-ccg.github.io/.well-known/did.json
+ -> https://w3c-ccg.github.io/did.json
```

Comment:
This is a breaking change to some deployments of did:web. Is this change really needed, or is this just beautification? .well-known is kind of an established pattern for things like this.

Reply:
Consensus is moving away from established patterns while also breaking backward compatibility? The .well-known is not even visible in DID Docs, VCs etc. Why is this change needed?

Reply:
> This is a breaking change to some deployments of did:web.

did:web is still an experimental specification in the CCG. Breaking changes are expected.

> Why is this change needed? 

To simplify the implementation of the specification and reduce the attack surface of the DID Method. It has become clear that the algorithms for mapping did:web URIs to file path locations are problematic for a number of reasons -- people want to not support content negotiation, people don't want to support .well-known if they can get away from it, not using regular paths was a mistake. Now that there are more people implementing did:web, it's becoming clear that mistakes were made that should be fixed before the rest of the world picks the technology up.

That said, if there are huge deployments of did:web that we don't know about, there is probably a way to inject a compatibility layer into the specification, but the bar for putting those sorts of hacks into the specification should be high.

Reply:
> > This is a breaking change to some deployments of did:web.
> 
> did:web is still an experimental specification in the CCG. Breaking changes are expected.

Fair enough.

> 
> > Why is this change needed?
> 
> To simplify the implementation of the specification and reduce the attack surface of the DID Method. It has become clear that the algorithms for mapping did:web URIs to file path locations are problematic for a number of reasons -- some people don't want to support content negotiation, some people don't want to support .well-known if they can get away from it, some think that not using regular paths was a mistake, etc. Now that there are more people implementing did:web, it's becoming clear that mistakes were made that should be fixed before the rest of the world picks the technology up.

Just wondering, why is .well-known increasing the attack surface?

> That said, if there are huge deployments of did:web that we don't know about, there is probably a way to inject a compatibility layer into the specification, but the bar for putting those sorts of hacks into the specification should be high.

I'm not aware of a huge number of did:webs in production.

Reply:
or in other words, why would removing .well-known decrease the attack surface?

Reply:
@awoie currently the logic can be implemented like this: https://github.com/transmute-industries/verifiable-data/blob/main/packages/did-web/src/index.js

```
const didToUrl = (did) => {
  const regex = new RegExp(
    `did:web:(?<origin>[a-zA-Z0-9/.\\-_]+)(?<path>[a-zA-Z0-9/.:\\-_]*)`
  );
  const match = did.match(regex);
  if (match.groups.path) {
    return `https://${match.groups.origin}${match.groups.path.replace(
      /:/g,
      "/"
    )}/did.json`;
  }
  return `https://${match.groups.origin}/.well-known/did.json`;
};
module.exports = { didToUrl };
```

By removing the other case, it would become:

```
const didToUrl = (did) => {
  const regex = new RegExp(
    `did:web:(?<origin>[a-zA-Z0-9/.\\-_]+)(?<path>[a-zA-Z0-9/.:\\-_]*)`
  );
  const match = did.match(regex);
  if (match.groups.path) {
    return `https://${match.groups.origin}${match.groups.path.replace(
      /:/g,
      "/"
    )}/did.json`;
  }
 throw new Error('Malformed did:web");
};
module.exports = { didToUrl };
```

Its one less case for developers to implement incorrectly, and one less registry with IANA for us to forget to update.

Reply:
> I'm not aware of a huge number of did:webs in production.

@tplooker how many did webs you got in production?

Reply:
Our domestic COVID pass solution uses did:web as the issuer. I can't state official numbers, but here's a [public article](https://www.stuff.co.nz/national/health/coronavirus/300476004/overseas-new-zealander-left-waiting-23-days-for-vaccine-pass) that can give you an idea of the rough number of passes issued with did:web. 

The issue we'd have with this change is that we're relying on long caching periods in our verifier app to make offline verification possible and this change could require everyone using this app to verify passes in NZ to update their verification app which could impact the ability to verify valid passes. Again I can't mention specifics on number of installs, but Google does publish on the play store for android there's 100,000+ installs of this app already.

This is just a rough estimate of the size we're operating on with did:web and we're using the .well-known endpoint which makes this change that much more problematic for us. Can we please not intentionally change this unless it's absolutely necessary?

Reply:
@OR13 I see, I'm also supportive of simplifying the spec but after hearing @kdenhartog feedback I think we would need good reasons to change it. I'm also interested in @mirceanis feedback since they also have an implementation.

Reply:
@OR13 @msporny after hearing @kdenhartog feedback, do you have some thoughts on how backward compatibility could be done after introducing a breaking change like this?

Reply:
I'm hearing @kdenhartog say that they have ONE did:web published (one issuer, NZ Govt). Each COVID passport doesn't have a unique did:web identifier associated with it, does it?

Reply:
We have one issuer, but in order to resolve the trust gap of knowing that the pass is issued by a trusted issuer we have 100,000+ links to that single issuer which would need to be updated. For us it's not only about the number published but also about the number of back links to that single did:web.

Reply:
I could give you a rewrite rule to maintain compatibility if a change is made to the spec.

Reply:
The issue isn't a technical issue. In our code it would probably be quite small to update it. The impact is a deployment problem that's going to leave many non technical people clueless to why they're unable to verify passes that worked yesterday. Furthermore, we don't have an excellent way to notify everyone in the country to say "update your apps because we wanted to change the path of a URL".

Reply:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/302

Reply:
> @msporny takes bets on not all did:web implementations following 302 redirects :)

Reply:
Internal rewrites do not change the apparent URL - the server just returns the specified data as though it were at the target URL. (The DID would appear at both the new and the old location.)

Redirects are also possible, Whether those are allowed or respected depends on the client.

Reply:
> The impact is a deployment problem that's going to leave many non technical people clueless to why they're unable to verify passes that worked yesterday.

What would cause them to be unable to verify passes that worked yesterday? Why can't you keep the `did:web` file in the same location as it existed yesterday AND serve it from the new location?

Reply:
Note that Apple devices are somewhat famous for creating problems with 30x responses. Verify the best practice if using a redirect - I think it was 307 the last time I checked.

(An internal rewrite is cleaner from a client perspective.)

Reply:
> Why can't you keep the `did:web` file in the same location as it existed yesterday AND serve it from the new location?

If we let webservers be web servers, this conflict is easily resolved.

For example, the rewrite rule below will cause nginx to return the same file for the two common URLs that have been discussed.

```
rewrite ^/.well-known/did.json$ /did.json last;
```

Reply:
oh duh! 🤦  Sometimes it's the simplest things that are forgotten about. I think the redirect solution could be a simple and low impact (we'd still have to update a few things) change that I think could work. Let me check with others about if they're alright with this. It's going to take some time to build the consensus though so please bare with me while I check with others.

Reply:
You can do a redirect (307) that instructs the client to request the revised URL. This will work or fail depending on the client.

The webserver can also be configured with an internal rewrite that returns the DID file on an alternate URL, so that it appears at both URLs. I would recommend the internal rewrite as it is less risky.